Convert launcher icons to cairo surfaces

vkareh opened the pull request:
This improves support for HiDPI by loading properly scaled surfaces for
launcher and drawer icons.

It also fixes the Show Desktop wncklet to show a surface icon. Other
wncklets have their icons determined by libwnck, so they remain as
pixbufs.

Fixes mate-desktop/mate-desktop/issues/314

raveit65 commented:
Wow, panel-launcher and show-desktop icons are really sharp now.
![panel-launcher](https://user-images.githubusercontent.com/961604/39373298-6b962fb2-4a47-11e8-927c-56a5cdedf618.png)
![show-desktop-windows-list](https://user-images.githubusercontent.com/961604/39373299-6bb452d0-4a47-11e8-8557-393acdb7b652.png)
So windows-list icons are handled by libwnck?

Well, the animation from panel-launchers are gone. I guess they come back when the build warning is fixed.
Icons in clock-applet and na-tray-applet are another story?

vkareh commented:
@raveit65 - I'll check the build warnings. Fedora uses the `--with-x` flag, which Ubuntu doesn't use, so that's probably why I didn't get your same warnings.

* window-list, window-selector, and workspace-switcher are all handled by libwnck. I can take a look at this library (will probably reach out to @muktupavels for help/guidance)

* notification-area seems to scale properly on my side, I think it just loads the icons directly from whichever app places an icon there. Let me know if it's different on your side.

* clock-applet has no icons on my install - am I missing any? I guess maybe the clock-face when you open the calendar/locations thing. I can look into this, but not a super high priority.

* fish uses a tiny png - I was able to scale it, but it looked pretty bad, and I wasn't able to get the fish animation to work properly, so I'll defer. It might just be better to use a larger high-quality png for this.

raveit65 commented:
I forgot to say , shudown/lock-screen/logout/kill-applets are looking good.
![applets](https://user-images.githubusercontent.com/961604/39377917-208fbf0a-4a57-11e8-9fc6-2669e39e6cd9.png)

muktupavels commented:
for notification area icons you might need to restart related applications.

raveit65 commented:
> for notification area icons you might need to restart related applications.

I did, btw. i restarted session, same result.
![bildschirmfoto zu 2018-04-27 19-52-52](https://user-images.githubusercontent.com/961604/39378774-a1192808-4a59-11e8-8060-daa696e82737.png)

raveit65 commented:
I found also some repeating warnings in xsession-error log, happens when panel starts with session.
```
(mate-panel:4245): GLib-GObject-CRITICAL **: 11:24:21.338: g_object_set: assertion 'G_IS_OBJECT (object)' failed

(mate-panel:4245): GLib-GObject-CRITICAL **: 11:24:30.560: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.565: gdk_pixbuf_scale_simple: assertion 'GDK_IS_PIXBUF (src)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.565: gdk_pixbuf_get_width: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.565: gdk_pixbuf_get_n_channels: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.565: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.565: gdk_pixbuf_get_width: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GLib-GObject-CRITICAL **: 11:24:30.565: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(mate-panel:4245): Gtk-WARNING **: 11:24:30.565: drawing failure for widget 'GtkWindow': invalid value (typically too big) for the size of the input (surface, pattern, etc.)

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.589: gdk_pixbuf_scale_simple: assertion 'GDK_IS_PIXBUF (src)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.589: gdk_pixbuf_get_width: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.589: gdk_pixbuf_get_n_channels: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.589: gdk_pixbuf_get_height: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GdkPixbuf-CRITICAL **: 11:24:30.589: gdk_pixbuf_get_width: assertion 'GDK_IS_PIXBUF (pixbuf)' failed

(mate-panel:4245): GLib-GObject-CRITICAL **: 11:24:30.589: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

(mate-panel:4245): Gtk-WARNING **: 11:24:30.589: drawing failure for widget 'GtkWindow': invalid value (typically too big) for the size of the input (surface, pattern, etc.)
```
......and so on. I think the g_object warnings were still there before.

lukefromdc commented:
I do not get these warnings with the panel from master

vkareh commented:
@raveit65 - I fixed the build warnings (I never even noticed that animation - I usually have compositing disabled). I also fixed the clock weather.

I wasn't able to do much with the notification area, as Ubuntu seems to mess with it (they use the indicator applet, or ayatana, not sure which). The only thing I was able to load in the na-tray was HexChat, but that looked correctly scaled. I'll load this up on f28 tomorrow and try to reproduce (I might just submit a different pull request for the na-tray...)

I also wasn't able to improve the wnck icons, no matter what I tried - I started looking at the code in libwnck and I have a feeling I can do stuff there to improve HiDPI support, but it feels out of my league for now :stuck_out_tongue:

alexarnaud commented:
Are we sure the new implementation doesn't break accessibility ?

Best regards,
Alex.

raveit65 commented:
> Are we sure the new implementation doesn't break accessibility ?

Did you noticed a regression?

raveit65 commented:
Thanks a lot for your amazing work.
- The sun in clock-applet, his popup and clock-location-tile looks real nice now :)
![clock-applet-hidpi](https://user-images.githubusercontent.com/961604/39468460-60a3f596-4d33-11e8-92c0-d50acbf6a1f4.jpg)

- And the animation of the launchers is back, which looks very clear with HIDPI.

- Build warnings are gone.

- It seems that application icons of previews in workspace-switcher are displaying fine,
......sometimes i am not really sure if they are scaled or not, but this is OK for me.

- I agree with you to use different PRs for na-tray, and windows-list (maybe).

- Fish applet has another problem with the timing of the single images from the icons with HIDPI.
This needs to be fixed first. Does anyone use this applet nowadays?
So we can ignore this for the moment, imo.

And I can build this commit in a SRPM, RPM for you for better testing in f28.
If you like i can give you small instructions to rebuild the SRPM after you changed a patch.
This should make it easier for you.

In general this PR is a huge improvement of the appearance of the MATE desktop if using HIDPI :)

raveit65 commented:
I forgot to say that all warnings in xsession-errors log about the panel are gone.

vkareh commented:
@alexarnaud - I don't think this would break accessibility. The basis of the change is to convert icons from `GdkPixbuf` to `cairo_surface_t` so that they render well on HiDPI displays. The icons remain as icons, but please test and let me know if you notice any adverse effects.

alexarnaud commented:
Le 01/05/2018 à 13:09, Victor Kareh a écrit :
> @alexarnaud <https://github.com/alexarnaud> - I don't think this would 
> break accessibility. The basis of the change is to convert icons from 
> |GdkPixbuf| to |cairo_surface_t| so that they render well on HiDPI 
> displays. The icons remain as icons, but please test and let me know if 
> you notice any adverse effects.
> 

I've made a quick test and it works as before with the Orca screen reader.

Best regards,
Alex.
